// @tinymce/moxiedoc/dist/lib/reporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.error = exports.warn = exports.info = exports.debug = exports.addHook = exports.setLevel = void 0;
const clc = require("cli-color");
let currentLevel = 3 /* WARN */;
const hooks = [];
const log = (level, message) => {
    switch (level) {
        case 1 /* DEBUG */:
            message = clc.magenta('Debug: ') + message;
            break;
        case 2 /* INFO */:
            message = clc.cyan('Info: ') + message;
            break;
        case 3 /* WARN */:
            message = clc.yellow('Warning: ') + message;
            break;
        case 4 /* ERROR */:
            message = clc.red('Error: ') + message;
            break;
    }
    hooks.forEach((hook) => {
        hook(level, message);
    });
    // eslint-disable-next-line no-console
    console.log(message);
};
const createLogFunction = (level) => (...args) => {
    if (level >= currentLevel) {
        log(level, args.join(' '));
    }
};
const setLevel = (level) => {
    currentLevel = level;
};
exports.setLevel = setLevel;
const addHook = (hook) => {
    hooks.push(hook);
};
exports.addHook = addHook;
const debug = createLogFunction(1 /* DEBUG */);
exports.debug = debug;
const info = createLogFunction(2 /* INFO */);
exports.info = info;
const warn = createLogFunction(3 /* WARN */);
exports.warn = warn;
const error = createLogFunction(4 /* ERROR */);
exports.error = error;
