// @ephox/swag/lib/main/ts/ast/js/Serializer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
exports.__esModule = true;
exports.serialize = void 0;
var escodegen = __importStar(require("escodegen"));
var escapeStr = function (str) {
    return str.replace(/[\u007f-\uffff]/g, function (ch) {
        var code = ch.charCodeAt(0).toString(16);
        if (code.length <= 2) {
            while (code.length < 2) {
                code = '0' + code;
            }
            return '\\x' + code;
        }
        else {
            while (code.length < 4) {
                code = '0' + code;
            }
            return '\\u' + code;
        }
    });
};
var serialize = function (node) {
    var code = escodegen.generate(node, {
        comment: true,
        format: {
            indent: {
                style: '  ',
                base: 0
            },
            preserveBlankLines: false,
            safeConcatenation: true
        }
    });
    return escapeStr(code.trim());
};
exports.serialize = serialize;
