// @ephox/swag/lib/main/ts/api/NodeResolve.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
exports.__esModule = true;
exports.nodeResolve = void 0;
var CachedFileSystem_1 = require("../fs/CachedFileSystem");
var Resolve_1 = require("../fs/Resolve");
var defaultOptions = {
    basedir: '.',
    prefixes: [],
    mappers: [],
    forceFlat: true
};
var nodeResolve = function (options) {
    if (options === void 0) { options = defaultOptions; }
    var combinedOptions = __assign(__assign({}, defaultOptions), options);
    var prefixes = (0, Resolve_1.resolvePrefixPaths)(combinedOptions.basedir, combinedOptions.prefixes);
    var fs = options.fileSystem ? options.fileSystem : (0, CachedFileSystem_1.getFileSystem)();
    return {
        name: 'swag-resolve',
        resolveId: (0, Resolve_1.resolveId)(fs, prefixes, combinedOptions.mappers, combinedOptions.forceFlat)
    };
};
exports.nodeResolve = nodeResolve;
