// @tinymce/eslint-plugin/dist/main/ts/configs/Editor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.editor = void 0;
const Standard_1 = require("./Standard");
exports.editor = {
    extends: Standard_1.standard.extends,
    plugins: Standard_1.standard.plugins,
    rules: Object.assign(Object.assign({}, Standard_1.standard.rules), { '@tinymce/no-direct-editor-events': 'error', '@tinymce/no-direct-editor-options': 'error', '@tinymce/prefer-mcagar-tiny-assertions': 'error', '@tinymce/prefer-mcagar-tiny-dom': 'error', '@typescript-eslint/explicit-module-boundary-types': ['error', { allowArgumentsExplicitlyTypedAsAny: true }], '@typescript-eslint/comma-dangle': ['error', {
                arrays: 'only-multiline',
                objects: 'only-multiline',
                imports: 'never',
                exports: 'never',
                functions: 'never',
                enums: 'never',
                generics: 'never',
                tuples: 'never'
            }], 'arrow-body-style': 'off', 'import/order': ['error', {
                'newlines-between': 'always',
                'groups': [
                    ['builtin', 'external'],
                    'internal',
                    ['parent', 'sibling', 'index']
                ],
                'alphabetize': {
                    order: 'asc',
                    caseInsensitive: true
                }
            }], 
        // Disabled since we're using the typescript-eslint rule
        'comma-dangle': 'off' }),
    settings: {
        'import/internal-regex': '^(ephox|tinymce|tiny-premium)/.*'
    }
};
