// @tinymce/eslint-plugin/dist/main/ts/rules/NoDirectImports.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.noDirectImports = void 0;
const ExtractUtils_1 = require("../utils/ExtractUtils");
const ImportUtils_1 = require("../utils/ImportUtils");
exports.noDirectImports = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Disallows deep lib imports directly to a file within a @ephox, @tiny or @tinymce package.'
        },
        messages: {
            noDirectImport: 'Direct import to {{ moduleSpecifier }} is forbidden.'
        }
    },
    create: (context) => ({
        ImportDeclaration: (node) => {
            const moduleSpecifier = (0, ExtractUtils_1.extractModuleSpecifier)(node);
            if ((0, ImportUtils_1.isInternalLibModule)(moduleSpecifier) || (0, ImportUtils_1.isInternalSrcModule)(moduleSpecifier)) {
                context.report({
                    node,
                    messageId: 'noDirectImport',
                    data: { moduleSpecifier },
                });
            }
        }
    })
};
