// @tinymce/eslint-plugin/dist/main/ts/rules/NoEnumsInExportSpecifier.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.noEnumsInExportSpecifier = void 0;
exports.noEnumsInExportSpecifier = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Disallows enums to be exported though export specifier at the end of the file.'
        },
        messages: {
            noEnumsInExportSpecifier: 'Exports of enums needs to be where it\'s being declared.'
        }
    },
    create: (context) => {
        const enumNames = {};
        return {
            TSEnumDeclaration: (n) => {
                const node = n;
                enumNames[node.id.name] = true;
            },
            ExportSpecifier: (node) => {
                if (node.type === 'ExportSpecifier') {
                    const name = node.local.name;
                    if (enumNames.hasOwnProperty(name) && enumNames[name]) {
                        context.report({ node, messageId: 'noEnumsInExportSpecifier' });
                    }
                }
            }
        };
    }
};
