// @tinymce/eslint-plugin/dist/main/ts/rules/NoMainModuleImports.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.noMainModuleImports = void 0;
const ImportUtils_1 = require("../utils/ImportUtils");
const PathUtils_1 = require("../utils/PathUtils");
const ExtractUtils_1 = require("../utils/ExtractUtils");
const mainPathValidator = ImportUtils_1.isMainImport;
const genericPathValidator = (path) => (0, ImportUtils_1.isMainImport)(path) && (0, ImportUtils_1.isInternalPathAlias)(path);
exports.noMainModuleImports = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Disallows importing things directly from the Main module within src/main.'
        },
        messages: {
            noMainImport: 'Direct import to Main module is forbidden.'
        }
    },
    create: (context) => {
        const filename = (0, PathUtils_1.normalizeFilePath)(context.getFilename());
        const validator = (0, PathUtils_1.isPathInMain)(filename) ? mainPathValidator : genericPathValidator;
        return {
            ImportDeclaration: (node) => {
                const moduleSpecifier = (0, ExtractUtils_1.extractModuleSpecifier)(node);
                if (validator(moduleSpecifier)) {
                    context.report({
                        node,
                        messageId: 'noMainImport',
                    });
                }
            }
        };
    }
};
