// @tinymce/eslint-plugin/dist/main/ts/rules/NoPathAliasImports.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.noPathAliasImports = void 0;
const ExtractUtils_1 = require("../utils/ExtractUtils");
const ImportUtils_1 = require("../utils/ImportUtils");
const PathUtils_1 = require("../utils/PathUtils");
exports.noPathAliasImports = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Disallows importing things using path aliases within src/main.'
        },
        messages: {
            noPathAliasImport: 'Imports with path aliases within src/main is forbidden.'
        }
    },
    create: (context) => {
        if ((0, PathUtils_1.isPathInMain)(context.getFilename())) {
            return {
                ImportDeclaration: (node) => {
                    if ((0, ImportUtils_1.isInternalPathAlias)((0, ExtractUtils_1.extractModuleSpecifier)(node))) {
                        context.report({
                            node,
                            messageId: 'noPathAliasImport',
                        });
                    }
                }
            };
        }
        else {
            return {};
        }
    }
};
