// @tinymce/eslint-plugin/dist/main/ts/rules/NoPublicApiModuleImports.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.noPublicApiModuleImports = void 0;
const ImportUtils_1 = require("../utils/ImportUtils");
const PathUtils_1 = require("../utils/PathUtils");
const ExtractUtils_1 = require("../utils/ExtractUtils");
const mainPathValidator = ImportUtils_1.isPublicApiImport;
const genericPathValidator = (path) => (0, ImportUtils_1.isPublicApiImport)(path) && (0, ImportUtils_1.isInternalPathAlias)(path);
exports.noPublicApiModuleImports = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Disallows importing things directly from the PublicApi module outside demos.'
        },
        messages: {
            noPublicApiImport: 'Direct import to PublicApi module is forbidden outside demos.'
        }
    },
    create: (context) => {
        const filename = (0, PathUtils_1.normalizeFilePath)(context.getFilename());
        // Demo files are the sole location we want to allow
        if ((0, PathUtils_1.isPathInDemo)(filename)) {
            return {};
        }
        else {
            const validator = (0, PathUtils_1.isPathInMain)(filename) ? mainPathValidator : genericPathValidator;
            return {
                ImportDeclaration: (node) => {
                    const moduleSpecifier = (0, ExtractUtils_1.extractModuleSpecifier)(node);
                    if (validator(moduleSpecifier)) {
                        context.report({
                            node,
                            messageId: 'noPublicApiImport'
                        });
                    }
                }
            };
        }
    }
};
